### DISCUSSION TITLE
How can I get the changed history of a file by `GITHUB API`?

### DISCUSSION THREAD

**Discussion by BodyCSoulN** [community/community#38462] [Apps, API and Webhooks] [answered]
How can I get the changed history of a file by `GITHUB API`?

just like `git log filename`

**Comment by LangLangBart** [comment, marked as answer]
### GraphQL API
Example: 
- repo: [cli/cli](https://github.com/cli/cli)
  - file: .[gitattributes](https://github.com/cli/cli/commits/trunk/.gitattributes)

#### GraphQL QUERY
```graphql
{
  repository(owner: "cli", name: "cli") {
    defaultBranchRef {
      name
      target {
        ... on Commit {
          history(path: ".gitattributes", first: 100) {
            nodes {
              author {
                name
              }
              abbreviatedOid
              committedDate
            }
          }
        }
      }
    }
  }
}
```
#### RESULT
```json
{
  "data": {
    "repository": {
      "defaultBranchRef": {
        "name": "trunk",
        "target": {
          "history": {
            "nodes": [
              {
                "author": {
                  "name": "Mislav Marohnić"
                },
                "abbreviatedOid": "7232bdd",
                "committedDate": "2020-01-29T15:13:07Z"
              },
              {
                "author": {
                  "name": "Mislav Marohnić"
                },
                "abbreviatedOid": "4343863",
                "committedDate": "2020-01-07T14:22:43Z"
              }
            ]
          }
        }
      }
    }
  }
}
```

--- 

### final thoughts
- for more than 100 commits on a file you will need `pagination`, see the video below or the GitHub Docs
  - Informative talk about `GraphQL`  by Rea Loretta (15/Dec/19) - [Advanced patterns for GitHub's GraphQL API](https://www.youtube.com/watch?v=i5pIszu9MeM)
- I have not investigated how this works with the [REST API](https://docs.github.com/en/rest) or if it is even possible.
  - if you find a way, please publish your solution

**Reply by ghostinhershell** [reply]
Hey @BodyCSoulN 👋 

If @LangLangBart's response answered your question do you mind marking his response as the answer to your question? Thanks!
